✨ NEW "Magic" module - allows to control efficient and easy sessions!

TurboAnonym opened the pull request:
**old idea:**
_the setup will create an executable linking to pycaw.tools.app_mixer.
you can launch this mixer by using pycaw-app-mixer or pip -m pycaw.tools.app_mixer
this tool allows to change the volume and view changes to sessions._

_This is what I was talking about ;)_

AndreMiras commented on setup.py:
```diff
@@ -12,11 +12,16 @@ def read(fname):
     'comtypes', 'enum34;python_version<"3.4"', 'psutil', 'future']
 setup(name='pycaw',
-      version='20190904',
+      version='20210501',
```
Ouch I haven't released for all this time!
I would not bump this as part as the PR.
And that's a good reminder for me to make a release soon actually :sweat_smile:

TurboAnonym replied:
haha 😂

yeah of course- with all these brand new features 🤩

by the way: do you think we could just drop the python2 support?
I see a lot of big packages doing that... wouldn't it be time for that?

TurboAnonym replied:
haha 😂

yeah of course- with all these brand new features 🤩

by the way: do you think we could just drop the python2 support?
I see a lot of big packages doing that... wouldn't it be time for that?

AndreMiras replied:
Yeah I've been wondering for Python 2 so I'm glad you ask.
Python 2 officially retired January last year I think https://pythonclock.org/
I think we can release this version to be the last to support Python 2 and then drop it straight after that.

TurboAnonym replied:
hmmm does this mean ALL the new features need to be py2 compatible 😅

AndreMiras replied:
That's a good question, but new examples are already not Python2 compatible right?
I'll try to release this week or weekend, hold your horses :smile:

TurboAnonym replied:
Honestly i have no clue what works in Python 2 or not 😇 (I have never used it ^^)
... besides the obvious like print() and f strings.

So i guess its not quite 100% compatible...

I would be in for #team_lazy and drop the py2 support now 😂 its 2021!
We could leave the legacy release on github or something...

AndreMiras commented on pycaw/tools/app_mixer/README.md:
```diff
@@ -0,0 +1,14 @@
+_Demo:_
+https://user-images.githubusercontent.com/63397666/116756672-c5cfee00-aa0c-11eb-9cb5-c1f04e60acfa.mp4
+
+## controll and view live audio sessions:
```
Typo s/controll/control/g Also in the commit body

TurboAnonym replied:
thanks for pointing that out ;)

AndreMiras commented on pycaw/tools/app_mixer/app_controll_and_gui.py:
```diff
@@ -0,0 +1,248 @@
+    handles user input and changing session volume/mute.
+    """
+
+    def __init__(self, RootFrameInstance, session, iid, session_manager):
```
If `RootFrameInstance` is really an object instance rather than a type/class, then why isn't not snake case notation?

TurboAnonym replied:
... something learned again ;)

AndreMiras commented on pycaw/tools/app_mixer/app_controll_and_gui.py:
```diff
@@ -0,0 +1,248 @@
+            self.app_name = "System Sounds"
+
+        # ______________ CREATE FRAME ______________
+        Frame.__init__(self, RootFrameInstance)
```
Just curious, why not `super().__init__(RootFrameInstance)`

TurboAnonym replied:
ahh i remember - it was copy pasted xD

AndreMiras commented on pycaw/tools/app_mixer/callbacks_and_manager.py:
```diff
@@ -0,0 +1,112 @@
+
+    @classmethod
+    def remove_session(cls, iid):
+        if iid in cls.active:
+            del cls.active[iid]
```
We could make this one-liner with:
```suggestion
        cls.active.pop(iid, None)
```

TurboAnonym replied:
hold on - genius move 😇

AndreMiras commented on pycaw/tools/app_mixer/callbacks_and_manager.py:
```diff
@@ -0,0 +1,112 @@
+            # ________ REMOVES 1 ITEM FROM LIST ________
+            _, app_row = cls.active.popitem()
+
+            try:
+                app_row.audio_session.unregister_notification()
+            except TypeError as e:
+                # should raise an error if a session is closed
+                # between popitem() and unregister_notification()
+                # but if you are that lucky - you should go to the lottery
+                print("you won the lottery")
+                print(str(e))
```
If at the end we do nothing with it how about ignoring with https://docs.python.org/3/library/contextlib.html#contextlib.suppress:
```suggestion
            with contextlib.suppress(TypeError):
                # could raise an error if a session is closed
                # between popitem() and unregister_notification()
                # but if you are that lucky - you should go to the lottery
                app_row.audio_session.unregister_notification()
```
But honestly I would not bother to even catch this error if it has little chance to occur

TurboAnonym replied:
yeah nice idea!

i will remove it here then ;)

AndreMiras commented on pycaw/tools/app_mixer/callbacks_and_manager.py:
```diff
@@ -0,0 +1,112 @@
+
+class AudioSessionCallback(AudioSessionEvents):
+    """Audio session callback"""
+    def __init__(self, AppRowInstance):
```
If `AppRowInstance` is an instance it should be snake case

TurboAnonym replied:
okay ^^

AndreMiras commented on pycaw/tools/app_mixer/launch.py:
```diff
@@ -0,0 +1,88 @@
+        ]
+
+        # add sessions to session_manager
+        [session_manager.add_session(session) for session in all_sessions]
```
I'd rather make it a tuple (immutable) specially if it's saved as a member
```suggestion
        tuple(session_manager.add_session(session) for session in all_sessions)
```

TurboAnonym replied:
👍
... but just out of curiosity: what do you mean with "if it's saved as a member"

AndreMiras replied:
My bad I miss read that line somehow sorry :facepalm: 
I read it as something like:
```python
session_manager = [session_manager.add_session(session) for session in all_sessions]
cls.session_manager = session_manager
```
Not sure how I brainfarted like this. You'were just doing a `add_session()` via a list comprehension.
Yeah you can revert it doesn't make sense what I wrote and it makes things confusing

TurboAnonym replied:
okay ;) no problem ;)

AndreMiras commented on pycaw/tools/app_mixer/launch.py:
```diff
@@ -0,0 +1,88 @@
+    StartUp(RootFrameInstance, AppRow)
+
+    # ______________ START WINDOW ______________
+    try:
+        RootFrameInstance.mainloop()
+    except KeyboardInterrupt:
+        pass
```
```suggestion
    with suppress(KeyboardInterrupt):
        RootFrameInstance.mainloop()
```

TurboAnonym replied:
🙌

AndreMiras commented on pycaw/tools/app_mixer/launch.py:
```diff
@@ -0,0 +1,88 @@
+    # ________________ CLEAN UP ________________
+    StartUp.unregister_all_notifications()
+
+    print("\nTschüss")
```
:thinking: :sweat_smile:

TurboAnonym replied:
yeah it doesnt seem to make much sense 😂

I will think of another class name.

AndreMiras reviewed (commented):
Yeah I know it's WIP, but I couldn't help it.
Looking good already!
Made some early comments :smiley:

TurboAnonym reviewed (commented):
Thank you (really!) for all the feedback!

It helps a lot improving my coding ^^

TurboAnonym commented:
I split the functionality into a new module named magic 😇

This is a simple example:

Magic:
------
The new Magic module allows to easy access and control the sessions of apps without headache!

example for MagicApp

```
from pycaw.magic import MagicApp

def changed(volume):
	print("Callback")
	print(volume)

m = MagicApp("msedge.exe", volume_callback=changed)

# start app for example now and play something!
# of course it would also work if app was already running.
time.sleep(6)

print(m.state)
print(m.volume)

m.volume = 0.6
m.mute = True
```

AndreMiras commented on pycaw/magic.py:
```diff
@@ -0,0 +1,563 @@
+"""
```
I'm sure you will write unit tests for this new module, right? :sweat_smile: :pray:

TurboAnonym replied:
hmm i started writing an example... that should do the trick in general right? ... of course not with msedge ... but similar

AndreMiras commented on pycaw/magic.py:
```diff
@@ -0,0 +1,563 @@
+
+        # normalize app_exe
+        if type(app_exe) not in (tuple, list):
+            app_exe = (app_exe,)
```
Curious, why not making it a set? It should be unique, plus the search would become ~O(1) rather than ~O(n)

TurboAnonym replied:
thats a great idea! 😄

AndreMiras commented on pycaw/magic.py:
```diff
@@ -0,0 +1,563 @@
+# TODO
+# i dont really now how to implement this.
+# this all should be actived once on MagicApp or MagicManager use
+# or already at import of this module?
```
Yeah I think ideally we should avoid magic to happen at import time for multiple reasons.
One is it makes it harder to (unit) test for instance

TurboAnonym replied:
yeah i intregated it with .activate_magic()

AndreMiras reviewed (commented):
Just gave it a first quick look even though it's WIP.
But maybe the magic module should come in another PR as current changes are big enough and already provide value.
What do you think?

TurboAnonym commented:
"But maybe the magic module should come in another PR as current changes are big enough and already provide value.
What do you think?"

its more like in reverse ... the initial idea for the pull was to create a pycaw-app-mixer but now it depends on the magic module...
(i splitted it up and added more functionality)

so i guess i will drop the pycaw-app-mixer for now and this pull will be for the magic module.

-------

I am contributing so much because i was searching for something like the magic module.
The closest to that was the initial pycaw. At first i didnt thought that a 'magic' way would ever exist to easily control the app volume - but here we are. The pycaw-app-mixer is more a fun by-product wich helped me understand windows audio sessions.

-------

My total goal is for the magic module to support:

for now:
- to control one or multiple sessions by app.exe (MagicApp)
- to register for each new session a wrapper to be able to do something like the pycaw-app-mixer (MagicSession)

later:
- to control the main speaker (not impl.)
- to control a custom device (also mics) by name (not impl.)

... and this all easily without (much) knowledge about: com (sta, mta) windows audio sessions and the callbacks and interfaces and so on.

TurboAnonym commented:
@AndreMiras can you please give me your opinion on the TODOS in the code? ;)

Heres a simplified sketch - how the magic module works.

![1Magic](https://user-images.githubusercontent.com/63397666/117002159-7d058700-ace3-11eb-8688-558ca82acbf0.png)

hight resolution as pdf:
[1Magic.pdf](https://github.com/AndreMiras/pycaw/files/6420990/1Magic.pdf)

AndreMiras commented on pycaw/magic.py:
```diff
@@ -0,0 +1,812 @@
+        # cls.empty_trash()
+
+        """
+        if cls.MagicSessionConfigured:
```
commented out code should be removed I guess, isn't it?

TurboAnonym replied:
https://github.com/AndreMiras/pycaw/pull/42#issuecomment-841659891

AndreMiras commented on pycaw/magic.py:
```diff
@@ -0,0 +1,812 @@
+    def mute(self, mute):
+        self.magic_root_session._sav.SetMute(mute, self.guid)
+
+    def toggle_mute(self):
```
Since `toggle_mute` and `step_volume` at least are used in two difference placed, maybe we could defined it as a function somewhere and call it from the method of both `MagicSession` and `MagicApp` what do you think?

TurboAnonym replied:
Let me introduce you to yet another parrent class ;)

AndreMiras commented on pycaw/magic.py:
```diff
@@ -0,0 +1,812 @@
+    # XXX remove old session:
+    # doesnt work this way either...
+    #
+    # def IUnknown_Release(self, *args, **kwargs):
```
I'd rather ditch any commented out code from the code base if possible

TurboAnonym replied:
https://github.com/AndreMiras/pycaw/pull/42#issuecomment-841659891

AndreMiras commented on pycaw/magic.py:
```diff
@@ -0,0 +1,812 @@
+
+    def OnStateChanged(self, new_state_id):
+        """Is fired, when the audio session state changed."""
+        if new_state_id == 2:
```
Could theses states mapped to some enum so it doesn't seems to be a magic value?

TurboAnonym replied:
Great ^^ 🙌👍

AndreMiras reviewed (commented):
Nice work, it looks pretty good.
I hope you don't mind I made few more comments.
Also I'd rather have the CI pass green before merging anything

TurboAnonym commented on pycaw/magic.py:
```diff
@@ -0,0 +1,812 @@
+            mgr = AudioUtilities.GetAudioSessionManager()
+            log.debug("<MagicManager/> got manager")
+        except COMError:
+            # exit("No speaker connected")
```
youre right ;) its time to clean up things ;)

TurboAnonym commented on examples/magic_app_example.py:
```diff
@@ -0,0 +1,85 @@
+# this import must happen before any other pycaw or comtypes import:
+from pycaw.magic import MagicApp
+
+"""non opp approach:"""
```
added ;)

TurboAnonym commented on pycaw/magic.py:
```diff
@@ -0,0 +1,812 @@
+        (sometimes a session fires multiple times OnSessionCreated
+        even its already there)
+        TODO: is that actually true? it didnt occur to me.
+        i guess i was wrong in the first place
```
Okay Im on on it ;)
i also tested pycaw.magic with 100 vlc media player audio sessions simultaneous ... all working correct - no duplicate.

TurboAnonym commented:
## There is now only one big problem left.
### ... And I cant really solve it - I bypassed it.

it has to do with all the
`# XXX remove old session:`
comments in the code.

### The problem is: How to really "remove" an expired session.

When a session makes the callback OnStateChanged with AudioSessionState.Expired it should be ideally removed afterwards, so no ressources get wasted. But I could only implement the first step: to session.unregister_notification()

I then tried to remove any references to the _MagicRootSession which would result in releasing connected COMObjects like ._ctl2 (the <POINTER(IAudioSessionControl2)/>).
As soon as I released the final reference from the IAudioSessionControl2 via a callback the program broke. Broke in terms of no further COM callbacks where received... setting the volume would still work. I think I encountered a deadlock ...

From my still little understanding of COM, it has to do with keeping references and dont release them during a callback.

As it is stated in https://docs.microsoft.com/en-us/windows/win32/api/audiopolicy/nn-audiopolicy-iaudiosessionevents:
- The client should never call the IAudioSessionControl::UnregisterAudioSessionNotification method during an event callback.
- The client should never release the final reference on a WASAPI object during an event callback.

more lecture here ...
https://docs.microsoft.com/en-us/windows/win32/api/audiopolicy/nf-audiopolicy-iaudiosessioncontrol-unregisteraudiosessionnotification

### bypass with a trash:

My idea was to implement a "trash" solution and keep in here references to deleted  magic_root_sessions.
But emptying the trash (and thus releasing <POINTER(IAudioSessionControl2)/>)  wont work in ANY callback, not even MagicManager.OnSessionCreated().

So in the current state the trash gets only emptied when the script is closed or when the user calls explicitly MagicManager.empty_trash()

I dont think that this is ideal. _But it works. (even with 150 trashed sessions)_


### More details:
You can easily deactivate my "trash" solution by commenting:
```
MagicManager.remove_session ->
cls.expired_magic_root_sessions.add(magic_root_session)
```

**Note:**
There is a slight difference in the callback behavior, when the "trash" solution is deactivated:
_When using MagicManager.magic_session(MagicSession) I enconter the deadlock after:_

close a session -> start a session -> deadlock 

_When using MagicApp mode or MagicManager.activate_magic() (without MagicSession)_

close a session -> deadlock

AndreMiras reviewed (approved):
Left one last minor comment even though I think it looks good enough for a merge.
And also another thing that's important to me is to have at least some basic tests for the magic.py module. Since this is now part of the library I think it should be tested, specially since it has quite some logic.
I understand it's not an easy one, so I'm happy to help in that and also do it in a follow up PR.
But I wouldn't usually merge code with that amount of logic and no unit tests

TurboAnonym commented on pycaw/magic.py:
```diff
@@ -0,0 +1,807 @@
+        return new_mute
+
+    def step_volume(self, step=0.1):
+        current = self.volume
+        if current is not None:
+            new = max(0, min(1, current + step))
+            if new != current:
+                self.volume = new
+                return new
+        return None
```
Thanks ;)

TurboAnonym commented:
Nice ;)
All done for this pull requests I guess ;)

I added with this push 'advanced_volume_callbacks', cleaned up some code comments and renamed some internal variables.
And made use of this 'guard clause pattern' thingy ... (thanks for the info ^^)

TurboAnonym commented:
One last thing:

i want to create a new issue regarding https://github.com/AndreMiras/pycaw/pull/42#issuecomment-841659891
but i feel like there a way to many opened - yet solved issues.
Can you please close the solved ones, so feature requests and new issues dont get lost?

AndreMiras reviewed (approved):
Thanks for the hard work and thanks for addressing all the comment.
I still have hope for unit tests in some (near?) future

TurboAnonym commented:
😉

https://github.com/AndreMiras/pycaw/pull/45

gitagogaming commented:
I do see its possible to get any potential application volumes, 

I was wondering if it is possible to get a callback on the default audio cable set inside of windows as well, so when the default audio levels change or mute I can know.

gitagogaming commented:
Also I've taken notice that Magic Session doesnt appear to be capturing all the proper processes?  
is this due to there not being any available callback from windows?

Is there some way for me to add a new session?  
This below allows me to get the proper sessions, but no callback is enabled.
```
global_states = []
sessions = AudioUtilities.GetAllSessions()
old_volume_list=[]
running = True
for s in sessions:
    if s not in global_states:
        try:
            global_states.append(s.Process.name())
        except AttributeError:
            pass
if old_volume_list != global_states:
    print( global_states)
    old_volume_list = global_states
